Vivek1/bankapp

vivekmenezes opened the pull request:
@petermattis here is the first version of the app. It actually doesn't work I suspect because of the incorrect use of transactions. I'll keep looking. Thanks

Commit:
First version of the banking demo app. It does the following:
1. Initializes all the bank accounts with a set balance
2. Runs a number of go routines that continuously transfer a random
amount between random bank accounts using transactions.
3. Checks that the total balance in the bank is unchanged.

This program fails at the moment perhaps because of my incorrect
use of transactions

Commit:
First version of the banking demo app. It does the following:
1. Initializes all the bank accounts with a set balance
2. Runs a number of go routines that continuously transfer a random
amount between random bank accounts using transactions.
3. Checks that the total balance in the bank is unchanged.

This program fails at the moment perhaps because of my incorrect
use of transactions

Commit:
:Merge branch 'vivek1/bankapp' of https://github.com/cockroachdb/cockroach into vivek1/bankapp

A few more comments

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+// Copyright 2014 The Cockroach Authors.
```
`s/2014/2015/g`

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+		log.Fatal(err)
+	}
+	kvClient := client.NewKV(nil, sender)
+	kvClient.User = storage.UserRoot
```
`client.KV` is thread-safe. You should create one and share it between all of the money transfer routines.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+		// Read to value.
+		call = client.Get(proto.Key(to))
+		gr = call.Reply.(*proto.GetResponse)
+		if err := txn.Run(call); err != nil {
```
You should read the from and to accounts in parallel. You do this by create 2 calls:

```
  get1 := client.Get(proto.Key(from))
  get2 := client.Get(proto.Key(to))
  if err := txn.Run(get1, get2); err != nil {
    return err
  }
```

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+		log.Fatal(err)
+	}
+
+	bank.approxNumTransactions++
```
There are atomic operations in the `sync/atomic` package.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+		log.Fatal(err)
+	}
+	kvClient := client.NewKV(nil, sender)
+	kvClient.User = storage.UserRoot
```
You should be sharing the same `kvClient` everywhere. It probably deserves to live in the `Bank` struct.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+
+// Bank stores all the bank related state.
+type Bank struct {
+	servingAddr string
```
I'd replace this with `kvClient`.

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+// makes an id string from an id int.
```
s/\/\/ ([a-z])/\/\/ \1/g

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+
+	putOpts := client.TransactionOptions{Name: "Initialize bank"}
+	for i := 0; i < numAccounts; i++ {
+		err = kvClient.RunTransaction(&putOpts, func(txn *client.Txn) error {
```
This initialization does not need to be performed in a transaction. It would be more efficient to not use one here. And you'd be better off run the puts in a batch. See my other comment about sending multiple calls to `Run`.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+	var bank Bank
+	bank.servingAddr = serv.ServingAddr()
+	bank.numAccounts = 100
+	const initCash int64 = 1000
```
In Go, you usually don'g have to give constants a type. It is fine to remove the `int64`. The constant will be untyped and will be converted to the correct type depending on context.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+	initBankAccounts(bank.servingAddr, bank.numAccounts, initCash)
+
+	// Start all the money transfer routines
+	const numTransferRoutines int = 10
```
`s/int//g`

petermattis commented:
I don't see anything obviously wrong with your use of transactions. What error are you seeing? It's possible its a bug in cockroach.

vivekmenezes commented:
Total cash in the bank = 100178

is what I'm seeing. It should be 100000

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+		for i := 0; i < bank.numAccounts; i++ {
+			call := client.Get(proto.Key(makeAccountID(i)))
+			gr := call.Reply.(*proto.GetResponse)
+			if err := txn.Run(call); err != nil {
```
Like elsewhere, you should perform all of the gets in parallel. Otherwise this is `O(num-accounts)` round-trips to the server.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,214 @@
+	// Start all the money transfer routines
+	const numTransferRoutines int = 10
+	for i := 0; i < numTransferRoutines; i++ {
+		go bank.continuousMoneyTransfer()
```
For debugging, I'd start by removing this `go` so that the money transfers happen synchronously. If the app still fails, then something basic is going wrong.

vivekmenezes replied:
Thanks, I figured it out. I was trying to move money within the same account thereby increase the value of some accounts. I'll brush up the PR and ping you for a review.

Commit:
:Merge branch 'vivek1/bankapp' of https://github.com/cockroachdb/cockroach into vivek1/bankapp

Incorporated all the feedback. parallelized execution of transactions whereever possible.

Commit:
Merge branch 'vivek1/bankapp' of https://github.com/cockroachdb/cockroach into vivek1/bankapp

Parallelized some of the transactions.

Commit:
Changed a few comments

vivekmenezes commented:
@petermattis ready for another viewing. Thanks.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,195 @@
+		log.Fatal(err)
+	}
+	atomic.AddInt32(&bank.numTransactions, 1)
+	return true
```
Our client API is too verbose and seems to obfuscate what this program is actually doing.

vivekmenezes replied:
I think there is value to having the []byte stored in the value be returned via an API to the user instead of the user needing to deal with the proto message stuff.

petermattis replied:
Yeah, it feels like the current API is just a straightforward exposure of the protocol instead of providing a more natural interface.

petermattis commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,195 @@
+		go bank.continuousMoneyTransfer()
+	}
+	// Sleep for a bit to allow money transfers to happen in the background and then exit.
+	time.Sleep(10 * time.Second)
```
Rather than checking the balances once, I think I'd put this in a loop where you sleep for a second then wake up and check the balances. That way the demo just runs until the user ctrl-c's it.

petermattis commented:
LGTM, but I think our client API needs some refinement.

Commit:
Add periodic bank account total  check.

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,200 @@
+			toValue = readValue
+		}
+		// Update both accounts.
+		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))))
```
You're deliberately not using increments, right?

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,200 @@
+// is available in the from account. Returns true on success.
+func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+	// Early exit when from == to
+	if bytes.Compare(from, to) == 0 {
```
`bytes.Equal` does the same and looks slightly better.

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,200 @@
+		}
+		// Update both accounts.
+		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))))
+		txn.Prepare(client.Put(proto.Key(to), []byte(fmt.Sprintf("%d", toValue+amount))))
```
being able to skip the `proto.Key` here would be good. And being able to pass string or byte in both slots as well. Could easily get that by passing interfaces and type assertions. Having to pass an error back from `Put` is awkward though, but in most puts you're always trying to pass the same thing so that erroring out completely would be an option.

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -0,0 +1,200 @@
+		}
+		// Read to value.
+		var toValue int64
+		if toResp.Value != nil && toResp.Value.Bytes != nil {
```
some useful sugar could be `if v := toResp.GetBytes(); v != nil { .. }`, for integers analogously `if i, ok := r.GetInt(); ok { .. }`
